# Содержание <a class="anchor" id="begining"></a>

<h1><span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Содержание-" data-toc-modified-id="Содержание--1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Содержание <a class="anchor" id="begining" rel="nofollow"></a></a></span></li><li><span><a href="#Анализ-БД-крупного-сервиса-для-чтения-книг-по-подписке" data-toc-modified-id="Анализ-БД-крупного-сервиса-для-чтения-книг-по-подписке-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Анализ БД крупного сервиса для чтения книг по подписке</a></span></li><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Задача-1.-Подсчет-книг,--вышедших-после-1-января-2000-года" data-toc-modified-id="Задача-1.-Подсчет-книг,--вышедших-после-1-января-2000-года-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Задача 1. Подсчет книг,  вышедших после 1 января 2000 года</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Задача-2.-Подсчет-количества-обзоров-и-средней-оценки-каждой-книги" data-toc-modified-id="Задача-2.-Подсчет-количества-обзоров-и-средней-оценки-каждой-книги-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Задача 2. Подсчет количества обзоров и средней оценки каждой книги</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Задача-3.-Определить-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Задача-3.-Определить-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Задача 3. Определить издательство, которое выпустило наибольшее число книг толще 50 страниц</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Задача-4.-Определить-автора-с-самой-высокой-средней-оценкой-книг-(среди-книг-с-50-и-более-оценками)" data-toc-modified-id="Задача-4.-Определить-автора-с-самой-высокой-средней-оценкой-книг-(среди-книг-с-50-и-более-оценками)-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Задача 4. Определить автора с самой высокой средней оценкой книг (среди книг с 50 и более оценками)</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Задача-5.-Посчитать-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Задача-5.-Посчитать-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Задача 5. Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-8.1"><span class="toc-item-num">8.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Анализ БД крупного сервиса для чтения книг по подписке

**Цель исследования** - проанализировать базу данных, чтобы помочь сформулировать ценностное предложение для нового продукта.


**Задачи:**

- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги расчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так можно исключить из анализа брошюры;
- Определить автора с самой высокой средней оценкой книг  (среди книг с 50 и более оценками);
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

![](https://raw.githubusercontent.com/sashasepp/da_projects/main/13_sql/db.png)

# Обзор данных 

[К содержанию](#begining)

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры подключения к БД

db_config = {'user': '*****', # имя пользователя
'pwd': '*****',, # пароль
'host': '*****',,
'port': '*****',, # порт подключения
'db': ''*****','} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], db_config['pwd'], db_config['host'], db_config['port'],
db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

Посмотрим какие таблицы и данные есть в нашем распоряжении

In [3]:
#запрос
query = '''
SELECT *
FROM books
LIMIT 5;
'''

# вывод таблицы books
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
#запрос
query = '''
SELECT *
FROM authors
LIMIT 5;
'''

# вывод таблицы authors
pd.io.sql.read_sql(query, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [5]:
#запрос
query = '''
SELECT *
FROM ratings
LIMIT 5;
'''

# вывод таблицы ratings
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [6]:
#запрос
query = '''
SELECT *
FROM reviews
LIMIT 5;
'''

# вывод таблицы reviews
pd.io.sql.read_sql(query, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [7]:
#запрос
query = '''
SELECT *
FROM publishers
LIMIT 5;
'''

# вывод таблицы publishers
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

**Описание данных**


**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

# Задача 1. Подсчет книг,  вышедших после 1 января 2000 года

[К содержанию](#begining)

In [13]:
#запрос
query = '''
SELECT COUNT(*)
FROM books
WHERE publication_date >= '2000-01-01';
'''

# вывод результата
number_of_books =  pd.io.sql.read_sql(query, con = engine).iloc[0, 0]
display(number_of_books)
print(f'После после 1 января 2000 года вышлa {number_of_books} книгa')

821

После после 1 января 2000 года вышлa 821 книгa


## Вывод

В нашем сервисе достаточное количество современной литературы для наших подписчиков - есть из чего выбирать

# Задача 2. Подсчет количества обзоров и средней оценки каждой книги

[К содержанию](#begining)

In [9]:
#запрос
query = '''
WITH 
rev AS(SELECT book_id,
       COUNT(review_id) AS reviews
       FROM reviews
       GROUP BY book_id),

rat AS (SELECT book_id,
        AVG(rating) AS av_rating
        FROM ratings
        GROUP BY book_id)


SELECT b.title AS book,
       rev.reviews,
       rat.av_rating
FROM books AS b
LEFT JOIN rev ON b.book_id = rev.book_id
LEFT JOIN rat ON b.book_id = rat.book_id;
'''

# вывод результата
books_rating_reviews =  pd.io.sql.read_sql(query, con = engine)
books_rating_reviews

book  reviews  av_rating
0            The Body in the Library (Miss Marple  #3)      2.0   4.500000
1                                            Galápagos      2.0   4.500000
2                             A Tree Grows in Brooklyn      5.0   4.250000
3    Undaunted Courage: The Pioneering First Missio...      2.0   4.000000
4                                          The Prophet      4.0   4.285714
..                                                 ...      ...        ...
995   The Cat in the Hat and Other Dr. Seuss Favorites      NaN   5.000000
996    Anne Rice's The Vampire Lestat: A Graphic Novel      NaN   3.666667
997                          Essential Tales and Poems      NaN   4.000000
998                               Leonardo's Notebooks      NaN   4.000000
999                            The Natural Way to Draw      NaN   3.000000

[1000 rows x 3 columns]

## Вывод

В нашем сервисе достаточное количество обзоров и оценок для того, чтобы нашим подписчикам было проше сделать свой выбор 

# Задача 3. Определить издательство, которое выпустило наибольшее число книг толще 50 страниц

[К содержанию](#begining)

In [10]:
#запрос
query = '''
WITH 
b AS(SELECT book_id,
            publisher_id
     FROM books
     WHERE num_pages > 50)


SELECT p.publisher,
       COUNT(b.book_id) AS num_books
FROM publishers AS p
LEFT JOIN b ON p.publisher_id = b.publisher_id
GROUP BY p.publisher_id
ORDER BY num_books DESC
LIMIT 1;
'''

# вывод результата
top_publisher =  pd.io.sql.read_sql(query, con = engine)
display(top_publisher)
print(f'Больше всего книг толще 50 стр. у издательства {top_publisher.iloc[0, 0]}, которое выпустило {top_publisher.iloc[0, 1]} книги')

publisher  num_books
0  Penguin Books         42

Больше всего книг толще 50 стр. у издательства Penguin Books, которое выпустило 42 книги


## Вывод

Наиболее плодотворное сотрудничество у нас налажено с издательством Penguin Books

# Задача 4. Определить автора с самой высокой средней оценкой книг (среди книг с 50 и более оценками)

[К содержанию](#begining)

In [11]:
#запрос
query = '''
WITH

rat AS(SELECT book_id,
              rating
     FROM ratings
     WHERE book_id IN (SELECT b.book_id
                       FROM books AS b
                       LEFT JOIN ratings AS r ON b.book_id = r.book_id
                       GROUP BY b.book_id
                       HAVING COUNT(r.rating_id) >= 50))


SELECT a.author,
       AVG(rat.rating) AS avg_rating
FROM rat
LEFT JOIN books AS b ON rat.book_id = b.book_id
LEFT JOIN authors AS a ON b.author_id = a.author_id
GROUP BY a.author_id
ORDER BY avg_rating DESC
LIMIT 1;
'''

# вывод результата
top_author =  pd.io.sql.read_sql(query, con = engine)
display(top_author)
print(f'Автор с самой высокой средней оценкой книг - {top_author.iloc[0, 0]}, средняя оценка ее книг: {top_author.iloc[0, 1]}')

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré, средняя оценка ее книг: 4.287096774193548


## Вывод

Видимо, человечество до сих пор не смогло придумать ничего более захватывающего внимание читателей, чем книги о Гарри Поттере

# Задача 5. Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок

[К содержанию](#begining)

In [12]:
#запрос
query = '''
WITH

rev AS(SELECT username,
              COUNT(review_id) AS reviews
     FROM reviews
     WHERE username IN (SELECT username
                       FROM ratings
                       GROUP BY username
                       HAVING COUNT(rating_id) > 50)
     GROUP BY username)


SELECT AVG(reviews) AS avg_reviews
FROM rev
'''

# вывод результата
avg_reviews =  pd.io.sql.read_sql(query, con = engine)
display(avg_reviews)
print(f'Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок: {avg_reviews.iloc[0, 0]}')

avg_reviews
0    24.333333

Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок: 24.333333333333332


## Вывод

Наши критики достаточно опытны в плане рекомендации контента, и подписчики при выборе книг могут смело опираться на их советы

# Общий вывод

[К содержанию](#begining)

По итогам анализа базы данных было установлено:

1. 821 книга из нашего каталога была выпущена после 2000 года. Таким образом, в нашем сервисе вполне достаточное количество современной литературы и нашим подписчикам точно есть из чего выбирать


2. После подсчета количества обзоров и средней оценки каждой книги можно утверждать, что в нашем сервисе достаточное количество обзоров и оценок для того, чтобы нашим подписчикам было проше сделать свой выбор


3. В нашем каталоге больше всего книг толще 50 стр. у издательства Penguin Books, которое выпустило 42 книги. С ними у нас налажено наиболее плодотворное сотрудничество


4. Автор с самой высокой средней оценкой книг - Джоан Роулинг (около 4,3), что позволяет говорить, что книги о Гарри Поттере по-прежнему одни из самых любимых у наших подписчиков



5. Наконец, среднее количество обзоров от пользователей, которые поставили больше 50 оценок: 24.3. Таким образом, наши критики достаточно опытны в плане рекомендации контента, и подписчики при выборе книг могут смело опираться на их советы